File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/status/StatusBarModelTest.kt
package com.tyhp.lang.status

import com.tyhp.lang.binary.BinaryStatus
import com.tyhp.lang.debug.ProxyRunState
import com.tyhp.lang.lsp.LspClientState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class StatusBarModelTest {
    @Test
    fun `healthy project and running LSP`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.RUNNING,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/usr/local/bin/tyhp",
            ),
        )
        assertEquals("Tyhp · demo · ready", view.text)
        assertFalse(view.error)
        assertFalse(view.warning)
        assertTrue(view.tooltip.contains("/usr/local/bin/tyhp"))
        assertTrue(view.tooltip.contains("XDebug proxy: stopped"))
    }

    @Test
    fun `binary missing uses CLI missing copy`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.ERROR,
                binaryStatus = BinaryStatus.MISSING,
                binaryMessage = "Tyhp CLI was not found.",
            ),
        )
        assertEquals("Tyhp · demo · CLI missing", view.text)
        assertTrue(view.error)
        assertTrue(view.tooltip.contains("Tyhp CLI was not found"))
    }

    @Test
    fun `LSP error with a healthy binary`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.ERROR,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/bin/tyhp",
            ),
        )
        assertEquals("Tyhp · demo · LSP error", view.text)
        assertTrue(view.error)
    }

    @Test
    fun `not in a Tyhp project is a warning when LSP is ready`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "not in a Tyhp project",
                lspState = LspClientState.RUNNING,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/bin/tyhp",
            ),
        )
        assertEquals("Tyhp · not in a Tyhp project · ready", view.text)
        assertTrue(view.warning)
        assertFalse(view.error)
    }

    @Test
    fun `starting shows starting copy`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.STARTING,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/bin/tyhp",
            ),
        )
        assertEquals("Tyhp · demo · starting", view.text)
    }

    @Test
    fun `stopped LSP is distinct from error`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.STOPPED,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/bin/tyhp",
            ),
        )
        assertEquals("Tyhp · demo · LSP stopped", view.text)
        assertFalse(view.error)
        assertFalse(view.warning)
    }

    @Test
    fun `invalid binary is shown as CLI missing`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.STOPPED,
                binaryStatus = BinaryStatus.INVALID,
                binaryMessage = "not a file",
            ),
        )
        assertEquals("Tyhp · demo · CLI missing", view.text)
        assertTrue(view.error)
        assertTrue(view.tooltip.contains("not a file"))
    }

    @Test
    fun `running proxy is appended to a healthy status`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.RUNNING,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/bin/tyhp",
                proxyState = ProxyRunState.RUNNING,
                proxyDetail = "IDE 9003 / XDebug 9004",
            ),
        )
        assertEquals("Tyhp · demo · ready · proxy", view.text)
        assertTrue(view.tooltip.contains("XDebug proxy: listening (IDE 9003 / XDebug 9004)"))
    }

    @Test
    fun `proxy error is a warning when LSP is healthy`() {
        val view = formatStatusBar(
            StatusBarInput(
                projectLabel = "demo",
                lspState = LspClientState.RUNNING,
                binaryStatus = BinaryStatus.OK,
                binaryPath = "/bin/tyhp",
                proxyState = ProxyRunState.ERROR,
            ),
        )
        assertEquals("Tyhp · demo · proxy error", view.text)
        assertTrue(view.warning)
        assertFalse(view.error)
    }

    @Test
    fun `status-bar proxy actions cover start stop restart`() {
        assertEquals(listOf("start"), proxyStatusActions(ProxyRunState.STOPPED))
        assertEquals(listOf("stop", "restart"), proxyStatusActions(ProxyRunState.RUNNING))
        assertEquals(listOf("stop", "restart"), proxyStatusActions(ProxyRunState.STARTING))
        assertEquals(listOf("start", "restart"), proxyStatusActions(ProxyRunState.ERROR))
    }
}

File: tyhp-lang/vscode/src/status/statusBarModel.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { formatStatusBar, proxyStatusActions } from "./statusBarModel";

test("healthy project + running LSP", () => {
    const view = formatStatusBar({
        projectLabel: "demo",
        lspState: "running",
        binaryStatus: "ok",
        binaryPath: "/usr/local/bin/tyhp",
    });
    assert.equal(view.text, "$(check) Tyhp · demo · ready");
    assert.equal(view.error, false);
    assert.equal(view.warning, false);
    assert.match(view.tooltip, /\/usr\/local\/bin\/tyhp/);
});

test("binary missing uses error background and CLI missing copy", () => {
    const view = formatStatusBar({
        projectLabel: "demo",
        lspState: "error",
        binaryStatus: "missing",
        binaryMessage: "Tyhp CLI was not found.",
    });
    assert.equal(view.text, "$(error) Tyhp · demo · CLI missing");
    assert.equal(view.error, true);
    assert.match(view.tooltip, /Tyhp CLI was not found/);
});

test("LSP error with a healthy binary", () => {
    const view = formatStatusBar({
        projectLabel: "demo",
        lspState: "error",
        binaryStatus: "ok",
        binaryPath: "/bin/tyhp",
    });
    assert.equal(view.text, "$(error) Tyhp · demo · LSP error");
    assert.equal(view.error, true);
});

test("not in a Tyhp project is a warning when LSP is ready", () => {
    const view = formatStatusBar({
        projectLabel: "not in a Tyhp project",
        lspState: "running",
        binaryStatus: "ok",
        binaryPath: "/bin/tyhp",
    });
    assert.equal(view.text, "$(warning) Tyhp · not in a Tyhp project · ready");
    assert.equal(view.warning, true);
    assert.equal(view.error, false);
});

test("starting shows a spinner", () => {
    const view = formatStatusBar({
        projectLabel: "demo",
        lspState: "starting",
        binaryStatus: "ok",
        binaryPath: "/bin/tyhp",
    });
    assert.equal(view.text, "$(sync~spin) Tyhp · demo · starting");
});

test("running proxy is appended to a healthy status", () => {
    const view = formatStatusBar({
        projectLabel: "demo",
        lspState: "running",
        binaryStatus: "ok",
        binaryPath: "/bin/tyhp",
        proxyState: "running",
        proxyDetail: "IDE 9003 / XDebug 9004",
    });
    assert.equal(view.text, "$(check) Tyhp · demo · ready · proxy");
    assert.match(view.tooltip, /XDebug proxy: listening \(IDE 9003 \/ XDebug 9004\)/);
});

test("proxy error is a warning when LSP is healthy", () => {
    const view = formatStatusBar({
        projectLabel: "demo",
        lspState: "running",
        binaryStatus: "ok",
        binaryPath: "/bin/tyhp",
        proxyState: "error",
    });
    assert.equal(view.text, "$(warning) Tyhp · demo · proxy error");
    assert.equal(view.warning, true);
    assert.equal(view.error, false);
});

test("status-bar proxy actions cover start/stop/restart", () => {
    assert.deepEqual(proxyStatusActions("stopped"), ["start"]);
    assert.deepEqual(proxyStatusActions("running"), ["stop", "restart"]);
    assert.deepEqual(proxyStatusActions("starting"), ["stop", "restart"]);
    assert.deepEqual(proxyStatusActions("error"), ["start", "restart"]);
});
